### PR TITLE
feat(metadata): ability to get artifacts location for Argo-Workflows v3.0+

### PR DESCRIPTION
**Pull request by Subreptivus**
Starting from [this](https://github.com/argoproj/argo-workflows/pull/4618) change, Argo-Workflows isn't providing full data (with `provider` and `bucket` properties) in `outputs` annotation by default.
With this change `metadata_writer` will combine artifact file location from the `outputs` annotation and `provider` and `bucket` data from `archiveLocation` property of the `template` annotation in case of missing data in `outputs` annotation.

### CODE REVIEW

File: backend/metadata_writer/src/metadata_writer.py [hunk 1]
```diff
@@ -98,15 +97,24 @@ def get_object_store_provider(endpoint: str) -> bool:
...
+def argo_artifact_to_uri(artifact: dict, skip_key=False) -> str:
     # s3 here means s3 compatible object storage. not AWS S3.
     if 's3' in artifact:
         s3_artifact = artifact['s3']
-        return '{provider}://{bucket}/{key}'.format(
-            provider=get_object_store_provider(s3_artifact['endpoint']),
-            bucket=s3_artifact.get('bucket', ''),
-            key=s3_artifact.get('key', ''),
-        )
+        if (s3_artifact.keys() >= {'endpoint', 'bucket'}):
```
Comment (Ark-kun) [inline backend/metadata_writer/src/metadata_writer.py:104]: This comparison looks fragile. Can we do this check in a more robust way?

File: backend/metadata_writer/src/metadata_writer.py [hunk 2]
```diff
@@ -325,8 +333,14 @@ def is_kfp_v2_pod(pod) -> bool:
 
                     output_artifacts = []
                     for name, art in argo_output_artifacts.items():
-                        artifact_uri = argo_artifact_to_uri(art)
-                        if not artifact_uri:
+                        artifact_uri_check = argo_artifact_to_uri(art)
```
Comment (Ark-kun) [inline backend/metadata_writer/src/metadata_writer.py:336]: Can we just call this `artifact_uri`?

File: backend/metadata_writer/src/metadata_writer.py [hunk 3]
```diff
@@ -325,8 +333,14 @@ def is_kfp_v2_pod(pod) -> bool:
...
+                        if artifact_uri_check:
+                            if re.search('(\W+)', artifact_uri_check).group(1) != '://':
```
Comment (Ark-kun) [inline backend/metadata_writer/src/metadata_writer.py:338]: Should this be something like `if re.match(r'^\W+://', artifact_uri):`?

File: backend/metadata_writer/src/metadata_writer.py [hunk 4]
```diff
@@ -325,8 +333,14 @@ def is_kfp_v2_pod(pod) -> bool:
...
+                                artifact_uri_wo_key = argo_artifact_to_uri(argo_template.get('archiveLocation', {}), True)
```
Comment (Ark-kun) [inline backend/metadata_writer/src/metadata_writer.py:339]: Maybe we can always pass `argo_template.get('archiveLocation', {})` to `argo_artifact_to_uri` and encapsulate this logic there?

**Comment by Ark-kun** [comment]
Thank you for this contribution.

File: backend/metadata_writer/src/metadata_writer.py (hunk 1 above)
Comment (Subreptivus) [reply backend/metadata_writer/src/metadata_writer.py:104]: Hm, and why does it look fragile to you?
Looks clean and short check of multiple keys in dict to me comparing to checking for subset or making unnecessary loops.

File: backend/metadata_writer/src/metadata_writer.py (hunk 2 above)
Comment (Subreptivus) [reply backend/metadata_writer/src/metadata_writer.py:336]: Yes we can, it's just a habit of not overwriting variables while making some checks.

File: backend/metadata_writer/src/metadata_writer.py (hunk 4 above)
Comment (Subreptivus) [reply backend/metadata_writer/src/metadata_writer.py:339]: The thing is, that despite the property name in `archiveLocation` they're storing the folder of artifact location but not the actual archive (file) location. For example in the `outputs` you will have file location like `"key":"artifacts/artifact-passing-lr6fg/artifact-passing-lr6fg-2908156709/hello-art.tgz"` and within `archiveLocation` you will have something like `"key":"artifacts/artifact-passing-lr6fg/artifact-passing-lr6fg-2908156709"`

**Review by Arhell — COMMENTED** [review]
/ok-to-test

**Comment by Sharathmk99** [comment]
I'm facing same problem. Using KF version 1.3. Any workaround for now?

**Pull request closed without merging**